KAFKA-6913: Add Connect converters and header converters for short, integer, long, float, and double (WIP)

rhauch opened the pull request:
*[KIP-305](https://cwiki.apache.org/confluence/display/KAFKA/KIP-305%3A+Add+Connect+primitive+number+converters) has been approved.*

Added converters and header converters for the primitive number types for which Kafka already had serializers and deserializers. All extend a common base class, `NumberConverter`, that encapsulates most of the shared functionality. Unit tests were added to check the basic functionality.

These classes are not used by any other Connect code, and must be explicitly used in Connect workers and connectors.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-6913: Add Connect converters and header converters for short, integer, long, float, and double

Added converters and header converters for the primitive number types for which Kafka already had serializers and deserializers.

rhauch commented:
@rayokota, @mageshn, and @wicknicks: would you mind reviewing this simple PR for KIP-305?

wicknicks commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverterConfig.java:
```diff
@@ -0,0 +1,41 @@
+    private final static ConfigDef CONFIG;
+
+    static {
+        CONFIG = ConverterConfig.newConfigDef();
```
this can be inlined with the declaration above.

rhauch replied:
ack.

wicknicks reviewed (commented):
Looks great, @rhauch. One small comment.

wicknicks commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Override
+    public ConfigDef config() {
+        return null;
```
should this be an empty config?

rhauch replied:
no, it should return `NumberConverterConfig.configDef()`. fixed.

rhauch committed:
KAFKA-6913: Corrected two minor issues found by code review

rhauch commented:
@wicknicks, thanks for reviewing. I've added a commit that addresses the issues you found, and would appreciate another pass. Thanks!

wicknicks commented on connect/api/src/test/java/org/apache/kafka/connect/storage/DoubleConverterTest.java:
```diff
@@ -0,0 +1,72 @@
+
+public class DoubleConverterTest {
+    private static final String TOPIC = "topic";
+    private static final Double SAMPLE_VALUE = 1234.31d;
```
nit: the `d` is not required.

wicknicks commented on connect/api/src/test/java/org/apache/kafka/connect/storage/ShortConverterTest.java:
```diff
@@ -0,0 +1,72 @@
+
+public class ShortConverterTest {
+    private static final String TOPIC = "topic";
+    private static final Short SAMPLE_VALUE = 123;
```
do you think we should add the values of MIN and MAX for each of the types here? To ensure that they also work fine with the changes?

wicknicks reviewed (commented):
@rhauch thanks for the changes. added couple of questions.

rhauch committed:
KAFKA-6913: Cleaned up, refactored, and added more functionality to number converter tests

Added a base class to centralize all of the common number converter test methods, and added ability to test converting multiple number values.

rhauch commented:
@wicknicks, thanks again. I've added another commit that refactors the tests to move all of the test methods into a common base class for the number converter tests, and the test classes for each of the specific number converters now extends this and only override methods to provide type-specific sample data, converter, serializer, and schema. Multiple samples can be provided and they're all tested, and all specific number converter tests include min and max values along with the simple value used before.

I believe that this address both of your comments. I'd appreciate another review. Thanks!

wicknicks commented:
Thanks for the updates, @rhauch! LGTM.

mageshn commented on connect/api/src/test/java/org/apache/kafka/connect/storage/NumberConverterTest.java:
```diff
@@ -0,0 +1,83 @@
+    public void testBytesNullToNumber() {
+        SchemaAndValue data = converter.toConnectData(TOPIC, null);
+        assertEquals(schema(), data.schema());
+        assertEquals(null, data.value());
```
nit : can use assertNull

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
```
nit : should this be final?

rhauch replied:
No, I don't think it should be. Why artificially constrain subclasses, especially if others wanted to create custom subclasses?

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+    protected abstract T cast(Object value);
+
+    @Override
+    public byte[] fromConnectData(String topic, Schema schema, Object value) {
```
nit : should this be final?

rhauch replied:
No, I don't think it should be. Why artificially constrain subclasses, especially if others wanted to create custom subclasses?

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public SchemaAndValue toConnectData(String topic, byte[] value) {
```
nit : should this be final?

rhauch replied:
No, I don't think it should be. Why artificially constrain subclasses, especially if others wanted to create custom subclasses?

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public byte[] fromConnectHeader(String topic, String headerKey, Schema schema, Object value) {
```
nit : should this be final?

rhauch replied:
No, I don't think it should be. Why artificially constrain subclasses, especially if others wanted to create custom subclasses?

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public SchemaAndValue toConnectHeader(String topic, String headerKey, byte[] value) {
```
nit : should this be final?

rhauch replied:
No, I don't think it should be. Why artificially constrain subclasses, especially if others wanted to create custom subclasses?

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+        configure(conf);
+    }
+
+    protected abstract T cast(Object value);
```
nit : should this be implemented in every subclass? why not just (T) value in the fromConnectData;

rhauch replied:
We can use that in the default method, but IMO we want to keep the cast method so that subclasses can override the behavior if necessary.

mageshn replied:
I'm not sure if we would ever have a case where cats needs to be overridden but that seems like a reasonable approach instead of having cast implemented in every converter.

mageshn commented on connect/api/src/test/java/org/apache/kafka/connect/storage/NumberConverterTest.java:
```diff
@@ -0,0 +1,83 @@
+    protected abstract Schema schema();
+
+    @Before
+    public void setup() {
```
minor nit : why not use @parameterized and make schema, converter, serializer, etc part of parameters?

rhauch replied:
That will be easier with JUnit 5, but IMO parameterized tests in JUnit 4 are way complex and error prone given that they deal with `Collection<Object>` for the parameters. It's far simpler to just have these methods that the subclasses may want to use anyway.

I think another viable and pragmatic approach would be to use a constructor to specify the parameters. I'm not sure that's cleaner, so I went with the current approach.

mageshn replied:
May be the Converter, Serializer and Schema can be injected through the constructor. May be doesn't make sense for the samples(). I'm not too opinionated on the approach here since we can't use Junit 5.

mageshn reviewed (commented):
@rhauch Had few minor comments and questions

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+ * <p>
+ * This implementation currently does nothing with the topic names or header names.
+ */
+public abstract class NumberConverter<T extends Number> implements Converter, HeaderConverter {
```
Is this actually supposed to be public? Or does it need to be public for testing or something?

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverterConfig.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.Map;
+
+/**
+ * Configuration options for {@link StringConverter} instances.
```
looks like copy/paste leftover reference to `StringConverter`.

rhauch replied:
ack

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/storage/NumberConverter.java:
```diff
@@ -0,0 +1,121 @@
+        assert this.deserializer != null;
+        assert this.typeName != null;
+        assert this.schema != null;
+        assert this.schema.isOptional();
```
Is there a reason to enforce this here? Probably doesn't matter much since we're only exposing Converters that will use optional schemas, but it doesn't seem like there's any actual need in the class to verify this and if you wanted a class that enforced non-null values (which seems like a useful constraint for keys), you wouldn't be able to build it on top of this class given this assertion.

rhauch replied:
The contract documented in the JavaDoc for `NumberConverter` stipulated that the schema should be optional to support nulls. However, if we hide that base class, the concrete subclasses will document that contract and use the appropriate schema, so there's no reason for the check here.

ewencp reviewed (commented):
@rhauch Just a few minor comments, otherwise this is looking pretty much ready to go.

rhauch committed:
KAFKA-6913: Additional changes per code reviews.

rhauch commented:
@mageshn, @ewencp: I've added a commit that address your comments/suggestions. @mageshn, I responded with counterpoints on a few suggestions, and no changes were made with those.

Thanks!

mageshn commented on connect/api/src/test/java/org/apache/kafka/connect/storage/NumberConverterTest.java:
```diff
@@ -0,0 +1,84 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public abstract class NumberConverterTest<T extends Number> {
```
minor : do we need a test for a non numeric value or in general unsupported value?

rhauch replied:
I didn't think it was important, since we're really just delegating all functionality to the specific serializer and deserializer. IMO, the important thing is to verify that we're configuring and using the serializers and deserializers correctly.

rhauch replied:
The serializer and deserializer tests in `SerializationTest` (in the client module) already include a number of those conditions.

mageshn replied:
Sounds good. Thanks

rhauch replied:
However, given that the NumberConverter methods have a try-catch block, we should check that logic. I'll add a test for an invalid value.

rhauch replied:
Done.

rhauch committed:
KAFKA-6913: Added unit tests to check for try-catch logic around use of (de)serializers

rhauch commented:
retest this please

rhauch commented:
(The previous failures were unrelated broker unit tests.)

mageshn reviewed (approved):
LGTM. Thanks @rhauch

ewencp reviewed (approved):
LGTM, merging to master for 2.0.0

tomekl007 commented:
Hello, @ewencp.
I see your comment that converters should be merged and released with 2.0.0
Unfortunately, those classes are not present in the artifact with that version:
https://mvnrepository.com/artifact/org.apache.kafka/connect-api/2.0.0
There are only two converters: StringConverter and JsonConverter. 
Where I should look for converters from the KIP:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-305%3A+Add+Connect+primitive+number+converters
?
